server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,217 @@
+(Header) and payloads (Payloads). The request type for this function includes a
+`wait_timeout` property indicating how long the client is willing to wait for a
+synchronous response once the request has been accepted by the system. This
+value can be set to zero indicating no such waiting should occur.  The
+`request_id` property uniquely identifies the logical request (_not_ the
+issuance of the request) and is intended for use as an idempotency token.
+
+The second function, `AwaitUpdateWorkflowResponse` is used when a client has
+already issued an `UpdateWorkflow` request and has the resulting `update_id`.
+That `update_id` can be provided along with a `wait_timeout` (same semantics as
+above) to re-attach to an in-flight or completed Update and obtain the
+associated UpdateResult in the form of an `UpdateWorkflowResponse`.
+
+
+```protobuf
+service WorkflowService {
+    rpc UpdateWorkflow(UpdateWorkflowRequest) returns (UpdateWorkflowResponse){}
+    rpc AwaitUpdateWorkflowResponse(AwaitUpdateWorkflowResponseRequest) returns (UpdateWorkflowResponse){}
+}
+
+message UpdateWorkflowRequest {
+    temporal.api.common.v1.Header header = 1;
+    string request_id = 2;
+    string namespace = 3;
+    temporal.api.common.v1.WorkflowExecution execution = 4;
+    string update_name = 5;
+    temporal.api.common.v1.Payloads input = 6;
+    google.protobuf.Duration wait_timeout = 7;
+}
+
+message UpdateWorkflowResponse {
+    UpdateResultType result_type = 1;
+    string update_id = 2;
+    temporal.api.common.v1.Payloads output = 3;
+}
+
+enum UpdateResultType {
+    UPDATE_RESULT_TYPE_UNSPECIFIED = 0;
+    UPDATE_RESULT_TYPE_SUCCESS = 1;
+    UPDATE_RESULT_TYPE_ERROR = 2;
+}
+
+message AwaitUpdateResponseRequest {
+    string update_id = 1; 
+    google.protobuf.Duration wait_timeout = 2;
+}
+```
+
+## Revisiting Use Cases With Propsed API
+
+### Short Update
+
+Client calls `UpdateWorkflowRequest` with a suitable `wait_timeout`.
+
+### Uncertain Duration Update
+
+Client calls `UpdateWorkflowRequest` with an acceptable `wait_timeout`. If the
+`result_type` of the returned `UpdateWorkflowResponse` is
+`UPDATE_RESULT_TYPE_UNSPECIFIED` then the client can infer that the Update is
+still runing and can use the returned `update_id` to obtain the UpdateResponse
+at a later time with `AwaitUpdateWorkflowResponse`. If the `result_type` is any
+other value then the response `output` can be used directly.
+
+### Known Long Update
+
+Client calls `UpdateWorkflowRequest` with a zero `wait_timeout` and uses the
+returned `update_id` to obtain the UpdateResponse at a later time with
+`AwaitWorkflowResponse`.
+
+### Invalid Update
+
+If the UpdateRequest is rejected by the workflow execution while the RPC client
+is still connected, a suitable gRPC error can be returned synchronously. Note
+however rejected UpdateRequests are _not_ recorded in workflow history.
+
+### "Failing" Update
+
+An Update that is passes verification by the workflow code and runs to
```

Comment:
Kinda related to my other comment - why does an update need to pass workflow code validation? Why does workflow code even need to exist for it? I can think of many cons, but what benefits are there of requiring workflow code run to validate a request inline? Do those benefits outweigh the ability to provide updates on a workflow not on a worker at the time?

Reply:
It's to avoid that problem where a bad (possibly unintentionally so) actor floods a workflow's history with invalid updates. The hard requirement we settled on is that invalid requirements (which is a determination that can only be made by workflow code) _must not_ consume history space.

Reply:
Do you think if we reimplemented signals from scratch again we'd require worker code to execute before it was in history for fear of this bad actor? What about a bad actor flooding my workflow with signals? Are we saying we wish we hadn't allowed signals to avoid inline worker validation?

Some other related questions:

* Can update handling have side effects?
* Can update validation have side effects?
* Can I have custom update validation?

Reply:
fwiw, I'm in agreement with Chad here ... worker code shouldn't be in the way of writing history under any circumstances.

Reply:
I think having the validation and defaulting to preventing growing history unnecessarily makes good sense.

And at first blush, it also seems valuable to have the ability to fire off updates as fast as possible. However, when you think about "what's the usecase for firing off a signal/update and wanting it to be in history ASAP?" I think it's not immediately clear what that's for. Either you don't care about a result, in which case just use the zero timeout and from a caller's perspective it looks the same, or you do, in which case waiting for the round trip was necessary anyway.  The only situation seemingly solved by the super-quick write is being able to get an ID back as fast as possible, but that pushes you into the "actually I do care about the result" situation :thinking:

Reply:
For me, I don't want my update request to fail anymore than I would for start, cancel, or signal when the worker is not processing the workflow atm. Maybe I care about the response maybe I don't. Separating the request of an update from its execution where the worker is only required for the latter is a reasonable separation there I think. I don't think this should be phrased like "well if you're gonna have to wait for response anyways, might as well have a two-phase validation portion too", but instead "why are we adding such extreme complexity for this validation use case"? Is our lack of prevalidation today a real problem?

I think once implementation time comes around and we see how the APIs manifest in SDKs for validating vs processing the update as separate things, the complexity will rise. But I will go with the team if we still want to have worker validation before history placement.

A few more questions (sorry, last two kinda unrelated):

1. What validation timeout is leveraged when there is no worker processing the workflow at that point? How long will it wait for a worker to appear to process just the validation?
2. An update execution can be retried, but can an update validation be retried if the worker doesn't respond?
3. Does every update have to be responded to before the workflow completes?
4. I have heard we have no process-update timeout, so the update can take many weeks to be responded to right?

Reply:
@paulnpdev what it comes down to is whether we are okay allowing invalid updates to consume space on the history. We allow it for signals but at the onsite the consensus was that invalid updates _must not_ eat space in the history. Are you willing to give ground on the history requirement? Or alternatively, do you see a way to identify invalid updates without running workflow code?

Reply:
Another question I just thought of:

5. Can I use workflow state in my update validator? What if that state is one value during validation and another value during execution? Are we are going to ask them to re-validate before processing because they can't trust what was validated at validation time is still the case at processing time?

Reply:
The ability to reject signals was requested multiple times by users. We cannot do it for signals, but I think supporting it for updates is important. 

"StartUpdate" can happen even if a worker is down. The update will be rejected when the worker is back.

Reply:
AFAIR (with limited knowledge of the server architecture) we also mentioned that it'd reduce latency if we didn't have to write to history the way signals must be written.

Reply:
Answers for @cretz questions

1. Same behavior as any other pending task waiting for a Worker
2. Yes, it can and should be replayed
3. Great question and not one which I/we had spend any time considering. My gut instinct is to say that yes, workflows _can_ complete with "running" Updates. But that's an early instinct and the question needs more scrutiny.
4. Correct
5. Yes but as discussed, no mutation of state. The expectation is that many Update implementations will be of the conditional variety, e.g. "Do X if and only if state Y still has value Z" but that needs to be implemented by the workflow code.

Reply:
> what it comes down to is whether we are okay allowing invalid updates to consume space on the history. We allow it for signals but at the onsite the consensus was that invalid updates must not eat space in the history. Are you willing to give ground on the history requirement? Or alternatively, do you see a way to identify invalid updates without running workflow code?
@mmcshane I agree.  I'm saying that there's a tradeoff here.  I consider this and ROI question. If allowing invalid updates to consume history space buys us enough (statistically) in terms of latency and/or throughput (and my intuitions says it does) then that's a cost I'm willing to pay, as long as we maintain correctness.

Reply:
@mmcshane - Thanks. We need to make clear to users that their validation code will need to be reevaluated at update time because validation time and update time can differ by a long time and workflow state may have changed since then. (or maybe we expect SDKs to run validation code twice or whatever...ug, I am not a fan of this special validation step at all)

Reply:
After further discussion, I have been told two important things:

    We will not require a worker be running a workflow to accept an update from a client
    We will not separate the update validation and update execution steps across a server call boundary, only worker-local separation if any

Given these, all my concerns go away (can resolve this conversation, leaving open for my response to be visible).

Reply:
Although most of the discussion here was around preventing execution history from flood of invalid updates.  But the main thing we need to make sure we can process the update with a single write during `happy path`.  The ability to dispatch the update without any write and only perform the actual write when update is successfully processed is very essential to update use cases which are typically low latency.  Our ability to skip the history write when update is rejected a nice side effect of the above behavior.